### PR TITLE
Annotating base classes

### PR DESCRIPTION
**Pull request by Kakarot-SSJ4**

### CODE REVIEW

**Comment by Kakarot-SSJ4** [comment]
@kelloggm @panacekcz

File: src/main/java/org/apache/commons/lang3/CharUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/CharUtils.java]: Instead of suppressing this warning, I would annotate `str` as `@MinLen(1) String`. This follows the principle of annotating normal behavior: https://checkerframework.org/manual/#annotate-normal-behavior

File: src/main/java/org/apache/commons/lang3/CharUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/CharUtils.java]: Shouldn't this be `@NonNegative char ch` then? And then there would be no need to suppress the warning.

File: src/main/java/org/apache/commons/lang3/ClassUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/ClassUtils.java]: I think I suggested this on your other pull request, but you should put an `@EnsuresMinLenIf` annotation on `StringUtils.isEmpty` to avoid these false positives.

File: src/main/java/org/apache/commons/lang3/Conversion.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/Conversion.java]: Doesn't this mean that `src` should be `@MinLen(1)` then?

File: src/main/java/org/apache/commons/lang3/Conversion.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/Conversion.java]: Doesn't this mean that `src` should be `@MinLen(1)` then?

File: src/main/java/org/apache/commons/lang3/Conversion.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/Conversion.java]: Doesn't this mean that `src` should be `@MinLen(1)` then?

File: src/main/java/org/apache/commons/lang3/LocaleUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/LocaleUtils.java]: Since that throws an illegal argument exception, I would annotate `str` as `@MinLen(3)`

File: src/main/java/org/apache/commons/lang3/RandomStringUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/RandomStringUtils.java]: I don't understand what the null array has to do with a warning suppression here. Can you explain this a little more?

File: src/main/java/org/apache/commons/lang3/RandomStringUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/RandomStringUtils.java]: Why does the `null` array being handled differently mean that a warning is issued? Shouldn't the spec be accurate regardless?

File: src/main/java/org/apache/commons/lang3/RandomUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/RandomUtils.java]: That means that `count`'s type should be `@NonNegative int`. See https://checkerframework.org/manual/#annotate-normal-behavior

File: src/main/java/org/apache/commons/lang3/RandomUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/RandomUtils.java]: Could you add an `@LessThan` annotation to express this?

File: src/main/java/org/apache/commons/lang3/StringUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/StringUtils.java]: I would annotate `searchList` and `replacementList` as `@SameLen` of each other following https://checkerframework.org/manual/#annotate-normal-behavior

**Review by kelloggm — CHANGES_REQUESTED** [review]
Looks good overall! There were a few place where I think the documentation needs to be improved, because I didn't understand why you had to suppress a warning. There were also a few places where your annotations permitted clearly exceptional behavior, and you could have avoided suppressing warnings by writing annotations that reflected how methods are intended to be used. Those concerns are both relatively minor, though - on the whole, this is in good shape!

File: src/main/java/org/apache/commons/lang3/ClassUtils.java
Comment (Kakarot-SSJ4) [reply src/main/java/org/apache/commons/lang3/ClassUtils.java]: I have already annotated StringUtils.isEmpty that way.

File: src/main/java/org/apache/commons/lang3/RandomStringUtils.java
Comment (Kakarot-SSJ4) [reply src/main/java/org/apache/commons/lang3/RandomStringUtils.java]: If the array passed is null, then the set of all chars is used.
The error is issued because if `null` is passed as the array parameter, the other two parameters which are supposed to be `@IndexFor` and `@IndexOrHigh` are not such.
How do you propose I tackle this? Should I suppress the warnings in the `random()` function that is annotated and remove the annotations?

File: src/main/java/org/apache/commons/lang3/ClassUtils.java
Comment (kelloggm) [reply src/main/java/org/apache/commons/lang3/ClassUtils.java]: Then why is this warning suppression needed?

File: src/main/java/org/apache/commons/lang3/RandomStringUtils.java
Comment (kelloggm) [reply src/main/java/org/apache/commons/lang3/RandomStringUtils.java]: I see: the joy of not having optional parameters.

I think you've annotated it correctly, but I would move this text:

> If the array passed is null, then the set of all chars is used.
> The error is issued because if null is passed as the array parameter, the other two parameters which are supposed to be `@IndexFor` and `@IndexOrHigh` are not such

into the comment on the code

File: src/main/java/org/apache/commons/lang3/ClassUtils.java
Comment (Kakarot-SSJ4) [reply src/main/java/org/apache/commons/lang3/ClassUtils.java]: I think it is issuing a warning in #1 because of #2, after the substring, the checker assumes className may be a null string .
```
while (className.charAt(0) == '[') { // #1
            className = className.substring(1); // #2
        }
```

Comment (kelloggm) [reply src/main/java/org/apache/commons/lang3/ClassUtils.java]: That doesn't make sense. The Index Checker has no concept of nullness, so it wouldn't assume that the string is null.

It (rightly) thinks that `className` could be empty when entering the loop. How do you know that isn't the case? That would be the part of your comment about "className cannot end with a '['" - which I think should be the whole justification.

**Comment by Kakarot-SSJ4** [comment]
@kelloggm @panacekcz Please review.

File: Stubs/ThreadGroup.astub
Comment (kelloggm) [inline Stubs/ThreadGroup.astub:11]: Is there a reason not to add these to the Index annotated JDK? They seem pretty obviously correct, and classes in `java.lang` are great targets for annotations in the Index JDK.

File: src/main/java/org/apache/commons/lang3/ArrayUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/ArrayUtils.java]: Why not annotate `copyArrayGrow1` as `@MinLen(1)`? That's exactly the invariant you're expressing here.

File: src/main/java/org/apache/commons/lang3/ArrayUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/ArrayUtils.java]: Shouldn't the return type of `indexOf` be `@GTENegativeOne @LTLengthOf("array")`? And shouldn't the value of `INDEX_NOT_FOUND` be `-1`?

Similar code to this using the JDK's `indexOf` methods (e.g. on `String`s) is handled correctly by the Index Checker. Why isn't this?

File: src/main/java/org/apache/commons/lang3/ArrayUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/ArrayUtils.java]: You had to suppress this warning many times. I'm surprised that the Index Checker doesn't handle this cast correctly, because it seems like a relatively easy false positive to fix.

Can you file a bug in the `typetools/checker-framework` issue tracker with an example of one of these casts failing (any of the copies of this `remove` method should be fine)?

Once you've done that, you can change the warning suppression justification to just a link to issue, and it will be easy to remove those warning suppressions once the issue is fixed.

File: src/main/java/org/apache/commons/lang3/ArrayUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/ArrayUtils.java]: This explanation is a bit unclear. I would write this as "`removals is the number of items removed, which cannot be greater than the length of the original array, so srcLength - removals is @NonNegative`"

File: src/main/java/org/apache/commons/lang3/ArrayUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/ArrayUtils.java]: Why not?

File: src/main/java/org/apache/commons/lang3/ArrayUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/ArrayUtils.java]: Should the type of this array be `final @IndexFor("array") int[] indices` - that is, an array of indices into `array`?

The justification for `#4` seems to indicate that this is the type.

The checker usually infers the types of local variables, but does not infer component types (the types of array elements, type parameters, etc), so you have to annotate them.

File: src/main/java/org/apache/commons/lang3/CharSequenceUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/CharSequenceUtils.java]: typo: `cs.length()`, not `cz.length()`

File: src/main/java/org/apache/commons/lang3/CharSequenceUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/CharSequenceUtils.java]: Just link the pull request

File: src/main/java/org/apache/commons/lang3/CharUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/CharUtils.java]: No need for this comment

File: src/main/java/org/apache/commons/lang3/CharUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/CharUtils.java:340]: Is this an instance of [typetools/checker-framework#2540](https://github.com/typetools/checker-framework/issues/2540) or something very similar? Another GSoC student is working on fixing that right now, so commenting on that issue with a test case with this method in it would be useful. Can you do that?

File: src/main/java/org/apache/commons/lang3/ClassUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/ClassUtils.java]: typo: "acording" -> "according"

File: src/main/java/org/apache/commons/lang3/RandomUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/RandomUtils.java]: Unneeded comment

File: src/main/java/org/apache/commons/lang3/StringUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/StringUtils.java]: typo: whatacter

File: src/main/java/org/apache/commons/lang3/text/translate/EntityArrays.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/text/translate/EntityArrays.java]: ?

**Review by kelloggm — CHANGES_REQUESTED** [review]
Overall, this looks really good. There are still some things that need to be looked at, especially things that I think ought to be annotated differently to remove a lot of suppressed warnings, and things that ought to be reported as bugs in the checker.

File: src/main/java/org/apache/commons/lang3/ArrayUtils.java
Comment (Kakarot-SSJ4) [reply src/main/java/org/apache/commons/lang3/ArrayUtils.java]: The following is the stub file for java.lang.Comparable:
package java.lang;
```
interface Comparable<T> {
    int compareTo(T arg0);
}
```

**Comment by Kakarot-SSJ4** [comment]
@kelloggm you may review again

File: src/main/java/org/apache/commons/lang3/ArrayUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/ArrayUtils.java:352]: The return type of `clone()` on an array object should be `@SameLen` of the original array. 

We can probably accomplish that with a `@PolySameLen` annotation on the parameter and the return type in the annotated JDK. Can you open an issue in the main repository for this, and then try to add the fix and see if it works? Thanks.

File: src/main/java/org/apache/commons/lang3/ArrayUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/ArrayUtils.java]: You can eliminate this warning by making the cast type (`T[]`) correct. Right now, that cast is interpreted by the typechecker as `(T @UnknownVal [])`, which is safe but loses the `@MinLen(1)` information we need. If you change the type being casted into to `T @MinLen(1) []`, then the cast will succeed (`copyArrayGrow1` returns `@MinLen(1)`, after all) and no warning will be issued later.

File: src/main/java/org/apache/commons/lang3/ArrayUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/ArrayUtils.java:6077]: Now that you've opened [typetools/checker-framework#2603](https://github.com/typetools/checker-framework/issues/2603), you should link that issue at each warning suppression that it causes.

File: src/main/java/org/apache/commons/lang3/ArrayUtils.java
Comment (kelloggm) [reply src/main/java/org/apache/commons/lang3/ArrayUtils.java]: Why do you need these annotations on the arguments to `compare`, anyway? They seem kind of meaningless, since they don't control the type of the return value or anything.

File: src/main/java/org/apache/commons/lang3/CharUtils.java
Comment (kelloggm) [reply src/main/java/org/apache/commons/lang3/CharUtils.java:340]: The fix for the issue I linked has been merged to master. Can you check if that fix resolves this error, and if not can you write this up as a test case and open an issue?

File: src/main/java/org/apache/commons/lang3/RandomUtils.java
Comment (kelloggm) [inline src/main/java/org/apache/commons/lang3/RandomUtils.java]: Why is this error issued at all? The checker should be able to figure out which of two constants is larger! So this looks like a bug that you should report.

**Review by kelloggm — APPROVED** [review]
A few more issues that can be reported, and then some minor cleaning. But mostly ready to go to @panacekcz when you've made the changes I suggested here.

**Comment by kelloggm** [comment]
@Kakarot-SSJ4 what's the status of this PR? Are you waiting on a review from @panacekcz?

**Comment by Kakarot-SSJ4** [comment]
@kelloggm Yes I'm waiting on a review from @panacekcz

**Comment by kelloggm** [comment]
@Kakarot-SSJ4 in the future you should only tag the reviewer you need, then.

File: src/main/java/org/apache/commons/lang3/Conversion.java
Comment (panacekcz) [inline src/main/java/org/apache/commons/lang3/Conversion.java]: `src` should probably be `@MinLen(1)`. Similarly for other methods in this file, which are specified to throw exception if `src` is empty.

File: src/main/java/org/apache/commons/lang3/ArrayUtils.java
Comment (panacekcz) [inline src/main/java/org/apache/commons/lang3/ArrayUtils.java]: Indentation.

File: src/main/java/org/apache/commons/lang3/ObjectUtils.java
Comment (panacekcz) [inline src/main/java/org/apache/commons/lang3/ObjectUtils.java:645]: What is the purpose of these annotations? They don't make sense to me.

File: src/main/java/org/apache/commons/lang3/RandomStringUtils.java
Comment (panacekcz) [inline src/main/java/org/apache/commons/lang3/RandomStringUtils.java]: `count` should be `@NonNegative`. Similarly for other methods in this class.

File: src/main/java/org/apache/commons/lang3/RandomUtils.java
Comment (panacekcz) [inline src/main/java/org/apache/commons/lang3/RandomUtils.java]: `startIndex` and the result should be `@NonNegative`. Similarly for the method below.

File: src/main/java/org/apache/commons/lang3/ObjectUtils.java
Comment (Kakarot-SSJ4) [reply src/main/java/org/apache/commons/lang3/ObjectUtils.java:645]: Because checker issues an error on `c1.compareTo(c2)` saying that these annotations are required.

**Comment by Kakarot-SSJ4** [comment]
@panacekcz You can re-review

**Pull request merged**
